### PR TITLE
Fix channel state code and add backoff code

### PR DESCRIPTION
**Pull request by murgatroid99**
This streamlines some of the channel state transition code and adds proper channel reconnection support with exponential backoff in the pure JS implementation, in accordance with the [Connectivity Semantics and API](https://github.com/grpc/grpc/blob/master/doc/connectivity-semantics-and-api.md) and [Connection Backoff Protocol](https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md) documents.

I removed the idle timeout code because it wasn't correct, and according to grpc/grpc#13212, the C core doesn't have that either.

### CODE REVIEW

File: packages/grpc-js-core/src/channel.ts [hunk 1]
```diff
@@ -62,49 +72,97 @@ export interface Channel extends EventEmitter {
...
+        this.currentBackoffDeadline = new Date(now.getTime() + this.currentBackoff + uniformRandom(-jitterMagnitude, jitterMagnitude));
+      }
+      this.startConnecting();
+      break;
+    case ConnectivityState.READY:
+      this.emit('connect');
+      break;
+    case ConnectivityState.TRANSIENT_FAILURE:
+      this.subChannel = null;
+      this.backoffTimerId = setTimeout(() => {
```
Comment (kjin) [inline packages/grpc-js-core/src/channel.ts]: `this.backoffTimerId` is never used. Should it be used to cancel the timeout on shutdown?

File: packages/grpc-js-core/src/channel.ts [hunk 2]
```diff
@@ -62,49 +72,97 @@ export interface Channel extends EventEmitter {
...
+        this.transitionToState([ConnectivityState.TRANSIENT_FAILURE], ConnectivityState.CONNECTING);
+      }, this.currentBackoffDeadline.getTime() - now.getTime());
+      break;
+    case ConnectivityState.IDLE:
+    case ConnectivityState.SHUTDOWN:
+      if (this.subChannel !== null) {
+        this.subChannel.shutdown({graceful: true});
```
Comment (kjin) [inline packages/grpc-js-core/src/channel.ts]: Looks like this is duplicated with https://github.com/grpc/grpc-node/pull/75/files#diff-6270a066e64c77ba494d58b714011deeR256 -- should the shutdown be removed from there?

File: packages/grpc-js-core/src/channel.ts (hunk 1 above)
Comment (murgatroid99) [reply packages/grpc-js-core/src/channel.ts]: yes, I've added that.

File: packages/grpc-js-core/src/channel.ts (hunk 2 above)
Comment (murgatroid99) [reply packages/grpc-js-core/src/channel.ts]: That's right. I removed the other instance.

**Pull request merged**